src/runtime/rpc/rpc_host_port.cpp

```diff
@@ -0,0 +1,190 @@
+            group_address->is_update_leader_automatically());
+        this->group_host_port()->set_leader(host_port(group_address->leader()));
+    } break;
+    default:
+        break;
+    }
+}
+
+void host_port::reset()
+{
+    switch (type()) {
+    case HOST_TYPE_IPV4:
+        _host.clear();
+        _port = 0;
+        break;
+    case HOST_TYPE_GROUP:
+        group_host_port()->release_ref();
+        break;
+    default:
+        break;
+    }
+    _type = HOST_TYPE_INVALID;
+}
+
+host_port &host_port::operator=(const host_port &other)
+{
+    if (this != &other) {
+        reset();
+        _type = other.type();
+        switch (type()) {
+        case HOST_TYPE_IPV4:
+            _host = other.host();
+            _port = other.port();
+            break;
+        case HOST_TYPE_GROUP:
+            _group_host_port = other.group_host_port();
+            group_host_port()->add_ref();
+            break;
+        default:
+            break;
+        }
+    }
+    return *this;
+}
+
+void host_port::assign_group(const char *name)
+{
+    reset();
+    _type = HOST_TYPE_GROUP;
+    _group_host_port = new rpc_group_host_port(name);
+    // take the lifetime of rpc_uri_address, release_ref when change value or call destructor
+    _group_host_port->add_ref();
+}
+
+std::string host_port::to_string() const
+{
+    switch (type()) {
+    case HOST_TYPE_IPV4:
+        return fmt::format("{}:{}", _host, _port);
+    case HOST_TYPE_GROUP:
+        return fmt::format("address group {}", group_host_port()->name());
+    default:
+        return "invalid address";
+    }
+}
+
+error_s host_port::resolve_addresses(std::vector<rpc_address> *addresses) const
+{
+    if (type() != HOST_TYPE_IPV4) {
+        return std::move(error_s::make(dsn::ERR_INVALID_STATE, "invalid host_port type"));
+    }
+
+    rpc_address rpc_addr;
+    if (rpc_addr.from_string_ipv4(this->to_string().c_str())) {
+        if (addresses) {
+            addresses->push_back(rpc_addr);
+        }
+        return error_s::ok();
+    }
+
+    struct addrinfo hints;
+    memset(&hints, 0, sizeof(hints));
+    hints.ai_family = AF_INET;
+    hints.ai_socktype = SOCK_STREAM;
+    AddrInfo result;
+    RETURN_NOT_OK(GetAddrInfo(_host, hints, &result));
+
+    // DNS may return the same host multiple times. We want to return only the unique
+    // addresses, but in the same order as DNS returned them. To do so, we keep track
+    // of the already-inserted elements in a set.
+    std::unordered_set<rpc_address> inserted;
+    std::vector<rpc_address> result_addresses;
+    for (const addrinfo *ai = result.get(); ai != nullptr; ai = ai->ai_next) {
+        CHECK_EQ(AF_INET, ai->ai_family);
+        sockaddr_in *addr = reinterpret_cast<sockaddr_in *>(ai->ai_addr);
+        addr->sin_port = htons(_port);
+        rpc_address rpc_addr(*addr);
+        LOG_INFO("resolved address {} for host_port {}", rpc_addr, to_string());
+        if (inserted.insert(rpc_addr).second) {
```

Comment:
This is to remove duplication, right? Can the input parameter `std::vector<rpc_address> *addresses` be of the std::set type?